Creating a Resource for an entry in a nested jar file in Jetty 12

**Jetty version**

Jetty 12.0.0.Beta2

**Java version**

17

**Question**

I'm working on upgrading Spring Boot to Jetty 12. Spring Boot supports executable jars and wars that can be run using `java -jar`. This results in URLs pointing to jars nested within a jar or war. For example, this URL taken from our test suite:

```
jar:file:/Users/awilkinson/dev/spring-projects/spring-boot/main/spring-boot-tests/spring-boot-integration-tests/spring-boot-server-tests/build/spring-boot-server-tests-app/build/libs/spring-boot-server-tests-app-jetty.war!/WEB-INF/lib/spring-boot-server-tests-app-resources.jar!/
```

Much of the path doesn't matter. A shorter example would be:

```
jar:file:/example.war!/WEB-INF/lib/spring-boot-server-tests-app-resources.jar!/
```

With Jetty 11 and earlier, I could create a `Resource` that points to the `META-INF/resources` entry within `spring-boot-server-tests-app-resources.jar!` by calling `Resource.newResource(url + "META-INF/resources")`. I am struggling to find the Jetty 12 equivalent of this code.

I have a `WebAppContext` to hand so I have tried calling `ResourceFactory resourceFactory = ResourceFactory.of(handler)` and calling `ResourceFactory.newResource`. This works for situations where `spring-boot-server-tests-app-resources.jar` is available directly on the file system but not for the case where it's nested within a war.

How can I get this working for nested jars using Jetty 12 please?

Thanks for the NPE fix. Unfortunately, it's not sufficient to reach parity with Jetty 11. Things now get a little further but the lack of support for detecting if the URL points to a directory still leaves us blocked. This is the failure that I'm seeing:

```
Caused by: java.lang.IllegalArgumentException: Non-Directory not allowed: URLResource@265FDED2(jar:file:/[…]/build/libs/spring-boot-server-tests-app-jetty.jar!/BOOT-INF/lib/spring-boot-server-tests-app-resources.jar!/META-INF/resources)
	at org.eclipse.jetty.util.resource.CombinedResource.gatherUniqueFlatResourceList(CombinedResource.java:87) ~[jetty-util-12.0.0-SNAPSHOT.jar!/:12.0.0-SNAPSHOT]
	at org.eclipse.jetty.util.resource.CombinedResource.combine(CombinedResource.java:44) ~[jetty-util-12.0.0-SNAPSHOT.jar!/:12.0.0-SNAPSHOT]
	at org.eclipse.jetty.util.resource.ResourceFactory.combine(ResourceFactory.java:46) ~[jetty-util-12.0.0-SNAPSHOT.jar!/:12.0.0-SNAPSHOT]
	at org.springframework.boot.web.embedded.jetty.JettyServletWebServerFactory.configureDocumentRoot(JettyServletWebServerFactory.java:327) ~[spring-boot-3.2.0-SNAPSHOT.jar!/:3.2.0-SNAPSHOT]
	... 19 common frames omitted
```

 If we overcome that, I suspect that listing not being supported will be the next problem.

--- a/jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResourceFactory.java
+++ b/jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResourceFactory.java
@@ -25,6 +25,7 @@ import java.nio.file.Path;
 import java.time.Instant;
 
 import org.eclipse.jetty.util.FileID;
+import org.eclipse.jetty.util.URIUtil;
 
 /**
  * {@link ResourceFactory} for {@link java.net.URL} based resources.
@@ -122,7 +123,7 @@ public class URLResourceFactory implements ResourceFactory
         @Override
         public URI getURI()
         {
-            return uri;
+            return URIUtil.correctFileURI(uri);
         }
 
         @Override
@@ -140,7 +141,7 @@ public class URLResourceFactory implements ResourceFactory
         @Override
         public Resource resolve(String subUriPath)
         {
-            URI newURI = uri.resolve(subUriPath);
+            URI newURI = resolve(uri, subUriPath);
             try
             {
                 return new URLResource(newURI, this.connectTimeout, this.useCaches);
@@ -148,6 +149,20 @@ public class URLResourceFactory implements ResourceFactory
             catch (MalformedURLException e)
             {
                 return null;
+            }
+        }
+
+        // This could probably live in URIUtil, but it's awefully specific to URLResourceFactory.
+        private static URI resolve(URI parent, String path)
+        {
+            if (parent.getPath() == null)
+            {
+                URI resolved = resolve(URI.create(parent.getRawSchemeSpecificPart()), path);
+                return URI.create(parent.getScheme() + ":" + resolved.toASCIIString());
+            }
+            else
+            {
+                return parent.resolve(path);
             }
         }
 

--- a/jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResourceFactory.java
+++ b/jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/URLResourceFactory.java
@@ -155,14 +155,19 @@ public class URLResourceFactory implements ResourceFactory
         // This could probably live in URIUtil, but it's awefully specific to URLResourceFactory.
         private static URI resolve(URI parent, String path)
         {
-            if (parent.getPath() == null)
+            if (parent.isOpaque() && parent.getPath() == null)
             {
                 URI resolved = resolve(URI.create(parent.getRawSchemeSpecificPart()), path);
                 return URI.create(parent.getScheme() + ":" + resolved.toASCIIString());
             }
-            else
+            else if (parent.getPath() != null)
             {
                 return parent.resolve(path);
+            }
+            else
+            {
+                // Not possible to use URLs that without a path in Jetty.
+                throw new RuntimeException("URL without path not supported by Jetty: " + parent);
             }
         }
 

--- a/jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/UrlResourceFactoryTest.java
+++ b/jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/resource/UrlResourceFactoryTest.java
@@ -15,6 +15,7 @@ package org.eclipse.jetty.util.resource;
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.net.MalformedURLException;
 import java.net.URI;
 import java.net.URL;
 import java.nio.ByteBuffer;
@@ -26,6 +27,7 @@ import java.time.Instant;
 
 import org.eclipse.jetty.toolchain.test.MavenTestingUtils;
 import org.eclipse.jetty.util.IO;
+import org.eclipse.jetty.util.URIUtil;
 import org.junit.jupiter.api.Tag;
 import org.junit.jupiter.api.Test;
 
@@ -119,6 +121,37 @@ public class UrlResourceFactoryTest
             int read = channel.read(buffer);
             assertThat(read, is(fileSize));
         }
+    }
+
+    @Test
+    public void testResolveUri() throws MalformedURLException
+    {
+        Path path = MavenTestingUtils.getTestResourcePath("example.jar");
+        URI jarFileUri = URI.create("jar:" + path.toUri().toASCIIString() + "!/WEB-INF/");
+
+        URLResourceFactory urlResourceFactory = new URLResourceFactory();
+        Resource resource = urlResourceFactory.newResource(jarFileUri.toURL());
+
+        Resource webResource = resource.resolve("web.xml");
+        assertThat(webResource.isDirectory(), is(false));
+        URI expectedURI = URI.create(jarFileUri.toASCIIString() + "web.xml");
+        assertThat(webResource.getURI(), is(expectedURI));
+    }
+
+    @Test
+    public void testResolveUriNoPath() throws MalformedURLException
+    {
+        Path path = MavenTestingUtils.getTestResourcePath("example.jar");
+        URI jarFileUri = URI.create("file:" + path.toUri().toASCIIString());
+
+        URLResourceFactory urlResourceFactory = new URLResourceFactory();
+        Resource resource = urlResourceFactory.newResource(jarFileUri.toURL());
+
+        Resource webResource = resource.resolve("web.xml");
+        assertThat(webResource.isDirectory(), is(false));
+
+        URI expectedURI = URIUtil.correctFileURI(URI.create("file:" + path.toUri().resolve("web.xml").toASCIIString()));
+        assertThat(webResource.getURI(), is(expectedURI));
     }
 
     private static long fileSize(URL url) throws IOException

